R/step_epi_slide.R

```diff
@@ -0,0 +1,277 @@
+step_epi_slide <-
+  function(recipe,
+           ...,
+           .f,
+           before = 0L,
+           after = 0L,
+           role = "predictor",
+           prefix = "epi_slide_",
+           f_name = clean_f_name(.f),
+           skip = FALSE,
+           id = rand_id("epi_slide")) {
+    if (!is_epi_recipe(recipe)) {
+      rlang::abort("This recipe step can only operate on an `epi_recipe`.")
+    }
+    .f <- validate_slide_fun(.f)
+    arg_is_scalar(before, after)
+    before <- try_period(before)
+    after <- try_period(after)
+    arg_is_chr_scalar(role, prefix, id)
+    arg_is_lgl_scalar(skip)
+
+    add_step(
+      recipe,
+      step_epi_slide_new(
+        terms = enquos(...),
+        before = before,
+        after = after,
+        .f = .f,
+        f_name = f_name,
+        role = role,
+        trained = FALSE,
+        prefix = prefix,
+        keys = epi_keys(recipe),
+        columns = NULL,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+
+step_epi_slide_new <-
+  function(terms,
+           before,
+           after,
+           .f,
+           f_name,
+           role,
+           trained,
+           prefix,
+           keys,
+           columns,
+           skip,
+           id) {
+    step(
+      subclass = "epi_slide",
+      terms = terms,
+      before = before,
+      after = after,
+      .f = .f,
+      f_name = f_name,
+      role = role,
+      trained = trained,
+      prefix = prefix,
+      keys = keys,
+      columns = columns,
+      skip = skip,
+      id = id
+    )
+  }
+
+
+
+#' @export
+prep.step_epi_slide <- function(x, training, info = NULL, ...) {
+  col_names <- recipes::recipes_eval_select(x$terms, data = training, info = info)
+
+  check_type(training[, col_names], types = c("double", "integer"))
+
+  step_epi_slide_new(
+    terms = x$terms,
+    before = x$before,
+    after = x$after,
+    .f = x$.f,
+    f_name = x$f_name,
+    role = x$role,
+    trained = TRUE,
+    prefix = x$prefix,
+    keys = x$keys,
+    columns = col_names,
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+
+
+#' @export
+bake.step_epi_slide <- function(object, new_data, ...) {
+  recipes::check_new_data(names(object$columns), object, new_data)
+  col_names <- object$columns
+  name_prefix <- paste0(object$prefix, object$f_name, "_")
+  newnames <- glue::glue("{name_prefix}{col_names}")
+  ## ensure no name clashes
+  new_data_names <- colnames(new_data)
+  intersection <- new_data_names %in% newnames
+  if (any(intersection)) {
+    nms <- new_data_names[intersection]
+    cli_abort(
+      c("Name collision occured. The following variable names already exist:",
+        `*` = "{.var {nms}}"
+      ),
+      call = caller_env()
+    )
+  }
+
+  ok <- object$keys
+  names(col_names) <- newnames
+  gr <- new_data %>%
+    dplyr::select(dplyr::all_of(c(ok, object$columns))) %>%
+    group_by(dplyr::across(dplyr::all_of(ok[-1]))) %>%
+    dplyr::arrange(time_value) %>%
```

Comment:
Many steps have a tendency to alter grouping and ordering. My convention is to always ensure that it happens inside the step to avoid bugs. But it would be nice to have an `epiprocess` function that uses canonical ordering cmu-delphi/epiprocess#482.